packages/react-dom/src/events/DOMModernPluginEventSystem.js

```diff
@@ -308,12 +326,62 @@ export function dispatchEventForPluginEventSystem(
   );
 }
 
+function getNearestRootOrPortalContainer(instance: Element): Element {
+  let node = getClosestInstanceFromNode(instance);
+  while (node !== null) {
+    const tag = node.tag;
+    // Once we encounter a host container or root container
+    // we can return their DOM instance.
+    if (tag === HostRoot || tag === HostPortal) {
+      return node.stateNode.containerInfo;
+    }
+    node = node.return;
+  }
+  return instance;
+}
+
 export function attachElementListener(listener: ReactDOMListener): void {
-  // TODO
+  const {event, target} = listener;
+  const {passive, type} = event;
+  let containerEventTarget = target;
+  // If we the target is a managed React element, then we need to
+  // find the nearest root/portal contained to attach the event listener
+  // to. If it's not managed, i.e. the window, then we just attach
+  // the listener to the target.
+  const possibleManagedTarget = ((target: any): Element);
+  if (getClosestInstanceFromNode(possibleManagedTarget)) {
+    containerEventTarget = getNearestRootOrPortalContainer(
+      possibleManagedTarget,
+    );
+  }
+  const listenerMap = getListenerMapForElement(containerEventTarget);
+  // Add the event listener to the target container (falling back to
+  // the target if we didn't find one).
+  listenToTopLevelEvent(
+    ((type: any): DOMTopLevelEventType),
+    containerEventTarget,
+    listenerMap,
+    passive,
+  );
+  // Get the internal listeners Set from the target instance.
+  let listeners = getListenersFromTarget(target);
+  // If we don't have any listeners, then we need to init them.
+  if (listeners === null) {
+    listeners = new Set();
+    initListenersSet(target, listeners);
+  }
+  // Finally, add our listener to the listeners Set.
+  listeners.add(listener);
 }
 
 export function detachElementListener(listener: ReactDOMListener): void {
-  // TODO
+  const {target} = listener;
+  // Get the internal listeners Set from the target instance.
+  const listeners = getListenersFromTarget(target);
+  if (listeners !== null) {
+    // Remove out listener from the listeners Set.
+    listeners.delete(listener);
+  }
```

Comment:
Does this method need to clean up anything in the listener map?

Reply:
In the follow up PR we will clear any leaking `listeners`. The listeners Set itself can be GC'd.

Reply:
It seems wrong that attaching adds things to map and set, but detaching only removes from set.

Reply:
Oh sorry, I missed your original point. Yep, I'll add clean-up here too.

Reply:
> The listeners Map isn't new too, it's been around for years – and we've never cleared it down.

Right. I guess I was assuming we'd need some check/logic to determine if this was the last thing listening to it, and we could then clear it out...but if we never did this, I guess there's no need to start doing it now. Just feels like a bug when I look at it.

> Lastly, it's not a bad case here – to over-listen.

I guess not, but at the same time having an *active* listener attached could have perf implications?

Ultimately I trust your judgement here b'c you're much more familiar with the surrounding code than I am. Just raising things that look potentially wrong as I see it :)

Reply:
It's valid to bring this up. I brought this up with @sebmarkbage many moons ago and he explained it to me and as to why we do it. Might be worth bringing it up again some time in a design meeting.